pkg/credentialprovider/azure/azure_credentials.go

```diff
@@ -186,24 +191,45 @@ func (a *acrProvider) Enabled() bool {
 }
 
 func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
-	klog.V(4).Infof("try to provide secret for image %s", image)
 	cfg := credentialprovider.DockerConfig{}
+	loginServer := a.parseACRLoginServerFromImage(image)
+	if loginServer == "" {
+		klog.V(4).Infof("image(%s) is not from ACR, return empty authentication", image)
+		return cfg
+	}
 
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// If the cache hasn't expired, return our cache
+	if time.Now().Before(a.expiration) {
+		if a.config != nil && a.config.UseManagedIdentityExtension {
+			// check whether loginServer is in cache
+			for k := range a.cacheDockerConfig {
+				if strings.EqualFold(loginServer, k) {
+					return a.cacheDockerConfig
```

Comment:
I still don't think this does what we want. If there are two subsequent calls to `Provide` with different login servers `foo1.azurecr.io` and `foo2.azurecr.io`, the second call will update `cacheDockerConfig` and invalidate caching for foo1.

Reply:
Agreed. Since this provider is caching on the registry level, its implementation should reflect that. This is essentially the request made in https://github.com/kubernetes/kubernetes/pull/92818#issuecomment-656261189 (to make the cache coherent with the significant portion of the image input).

It looks like the registry is the significant portion of the image input that affects the computed results, so the cache should be keyed by the registry when UseManagedIdentityExtension is enabled:

```
// cache credentials globally
cacheKey := ""
if a.config != nil && a.config.UseManagedIdentityExtension {
  // cache credentials at the registry level
  cacheKey = strings.Split(image, "/")[0]
}

cachedCredentials, found := cache.getUnexpiredCredentials(cacheKey)
if found {
  return cachedCredentials
}

... get credentials and store in cache under cacheKey ...
```

Reply:
@andrewsykim thanks, I have fixed it, see L213-L215

Reply:
@andrewsykim @liggitt could you take a look again? thanks
here `loginServer` is actually `cacheKey`, it's actually identical though with a little different implementation.

Reply:
> here `loginServer` is actually `cacheKey`, it's actually identical though with a little different implementation.
> `cacheDockerConfig[loginServer] = credential`

It would be helpful to make it clear that this logic can fit within the externalized credential provider API to compute the cache key the same way the provider will (taking the registry portion of the image, and not adding in case-insensitive comparisons)

Reply:
I have removed the case-insensitive comparisons, should fit within externalized credential provider API now. Let's merge this PR to get in-tree code work first, thanks.  @liggitt